src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1027,76 +943,116 @@
-                                                 (vhashes-to-ids %2))
-                                        pathstrs vhashes))
-         current-pairs (set (select-pid-vid-pairs-for-factset factset-id))
-         [new-pairs rm-pairs] (data/diff replacement-pv-pairs current-pairs)]
-
-     ;; Paths are unique per factset so we can delete solely based on pid.
-     (when rm-pairs
-       (let [rm-pids (set (map first rm-pairs))]
-         (jdbc/do-prepared
-          (format "DELETE FROM facts WHERE factset_id = ? AND fact_path_id %s"
-                  (jdbc/in-clause rm-pids))
-          (cons factset-id rm-pids))))
-
-     (insert-facts-pv-pairs! factset-id new-pairs)
-
-     (when rm-pairs
-       (delete-pending-path-id-orphans! factset-id
-                                        (set/difference
-                                         (set (map first rm-pairs))
-                                         (set (map first new-pairs))))
-       (delete-pending-value-id-orphans! factset-id rm-pairs))
-
-     (jdbc/update! :factsets
-                   {:timestamp (to-timestamp timestamp)
-                    :environment_id (ensure-environment environment)
-                    :producer_timestamp (to-timestamp producer_timestamp)
-                    :hash (-> fact-data
-                              shash/fact-identity-hash
-                              sutils/munge-hash-for-storage)}
-                   ["id=?" factset-id]))))
+          paths-and-valuemaps (facts/facts->paths-and-valuemaps values)
+          pathstrs (map #(facts/factpath-to-string (first %))
+                        paths-and-valuemaps)
+          pathstr->valuemap (into {} (map  (fn [pathstr [_ vm]] [pathstr vm])
+                                           pathstrs
+                                           paths-and-valuemaps))
+          existing-status ;; :changed :unchanged :gone
+          (jdbc/call-with-query-rows
+           [(str "select"
+                 "    path, id, value_type_id, large_value_hash,"
+                 "    case when large_value_hash is null then value end as value"
+                 "  from facts as f"
+                 "  inner join fact_paths fp on f.fact_path_id = fp.id"
+                 "  where f.factset_id = ?")
+            factset-id]
+           (fn [rows]
+             (doall
+              (for [{:keys [path id] :as row} rows]
+                (if-let [new-vm (pathstr->valuemap path)]
+                  (if (fact-value-changed? row new-vm)
+                    [:changed path id new-vm]
+                    [:unchanged path id])
+                  [:gone path id])))))
+
+          rm-pids (seq (map (fn [[_ _ pid]] pid)
+                            (filter #(= :gone (first %))
+                                    existing-status)))
+          new-pathstrs (let [have? (set (for [[st path] existing-status
+                                              :when #({:changed :unchanged} st)]
```

Comment:
looking a little closer it feels like this could actually be intended behavior, since we don't need to realize paths for unchanged facts? if that's the case it would be clearer to make it
```
:when #(= :changed st)
```